.github/workflows/lint-build-commits.yaml

```diff
@@ -36,41 +36,97 @@ jobs:
+        with:
+          path: /tmp/.cache/go
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+
+      # Load CCache build from GitHub
+      - name: Load ccache cache build from GitHub
+        uses: actions/cache@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
+        id: ccache-cache
+        with:
+          path: /tmp/.cache/ccache
+          key: ${{ runner.os }}-ccache-${{ hashFiles('bpf/**') }}
+          restore-keys: |
+            ${{ runner.os }}-ccache-
+
+      - name: Create cache directories if they don't exist
+        if: ${{ steps.go-cache.outputs.cache-hit != 'true' || steps.ccache-cache.outputs.cache-hit != 'true' }}
+        shell: bash
+        run: |
+          mkdir -p /tmp/.cache/go/.cache/go-build
+          mkdir -p /tmp/.cache/go/pkg
+          mkdir -p /tmp/.cache/ccache/.ccache
+
       - name: Check if build works for every commit
+        env:
+          CLANG: "ccache clang"
+          BUILDER_GOCACHE_DIR: "/tmp/.cache/go/.cache/go-build"
+          BUILDER_GOMODCACHE_DIR: "/tmp/.cache/go/pkg"
+          BUILDER_CCACHE_DIR: "/tmp/.cache/ccache/.ccache"
         run: |
           set -eu -o pipefail
           COMMITS=$(git rev-list ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }})
           for commit in $COMMITS ; do
             git checkout $commit || exit 1
-            contrib/scripts/builder.sh make build -j $(nproc) || exit 1
+            contrib/scripts/builder.sh make CLANG="${CLANG}" build -j $(nproc) || exit 1
           done
 
       - name: Check bpf code changes
         uses: dorny/paths-filter@de90cc6fb38fc0963ad72b210f1f284cd68cea36 # v3.0.2
         id: bpf-tree
         with:
+          # If these filters are modified, also modify the step below where we
+          # do a git diff
           filters: |
             src:
               - 'bpf/**'
 
       # Runs only if code under bpf/ is changed.
       - name: Check if datapath build works for every commit
         if: steps.bpf-tree.outputs.src == 'true'
+        env:
+          CLANG: "ccache clang"
+          BUILDER_GOCACHE_DIR: "/tmp/.cache/go/.cache/go-build"
+          BUILDER_GOMODCACHE_DIR: "/tmp/.cache/go/pkg"
+          BUILDER_CCACHE_DIR: "/tmp/.cache/ccache/.ccache"
         run: |
           set -eu -o pipefail
           COMMITS=$(git rev-list ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }})
           for commit in $COMMITS ; do
             git checkout $commit || exit 1
-            contrib/scripts/builder.sh make -C bpf build_all -j $(nproc) || exit 1
+            # Do not run make if there aren't any files modified in these
+            # directories from the previous commit to the current commit.
+            # If these filters are modified, also modify the step above where we
+            # run dorny/paths-filter.
+            if ! git diff --quiet HEAD^ bpf/ ; then
```

Comment:
This is correct as long as nothing in `bpf/` references things outside of the directory. But already the Makefile references `../Makefile.defs`. A change there could break the build, but we wouldn't notice it since it would be skipped here.

Reply:
If there are changes done in the Makefile, the if statement on line 97 would already prevent this step from running it, no?

Reply:
So what I have in mind is that you can have, say, three commits

1 change in bpf/blub
2 change in Makefile.defs (breaks the build)
3 change in bpf/blub (fixes the build)

Since there's changes in bpf the step is run, but since we're skipping the build after 2 we miss a broken build

Reply:
(discussed out of band):
 this is a best effort workflow, we won't be able to cover every dependency in the build - this is good enough for the common case, and crucially faster than what we had.